File: 77. Combinations/77. Combinations.rb
# @param {Integer} n
# @param {Integer} k
# @return {Integer[][]}
def combine(n, k)
    result = []
    backtrack(n, k, 1, [], result)
    result
end

def backtrack(n, k, start, current, result)
    if current.length == k
        result << current.dup
        return
    end
    
    (start..n).each do |i|
        current << i
        backtrack(n, k, i + 1, current, result)
        current.pop
    end
end

File: 77. Combinations/77. Combinations.go
func combine(n int, k int) [][]int {
    result := [][]int{}
    current := []int{}
    
    var backtrack func(int)
    backtrack = func(start int) {
        if len(current) == k {
            combination := make([]int, k)
            copy(combination, current)
            result = append(result, combination)
            return
        }
        
        for i := start; i <= n; i++ {
            current = append(current, i)
            backtrack(i + 1)
            current = current[:len(current)-1]
        }
    }
    
    backtrack(1)
    return result
}